enginetest/queries/join_queries.go

```diff
@@ -745,6 +745,328 @@ var JoinScriptTests = []ScriptTest{
+					{5, 50},
+				},
+			},
+			{
+				Query: "select * from t1 natural right join t2;",
+				Expected: []sql.Row{
+					{1, 30},
+					{2, 20},
+					{5, 50},
+				},
+			},
+			{
+				Query: "select t1.i, t1.j, t2.i, t2.j from t1 right join t2 using (i, j);",
+				Expected: []sql.Row{
+					{nil, nil, 1, 30},
+					{2, 20, 2, 20},
+					{nil, nil, 5, 50},
+				},
+			},
+			{
+				Skip:  true, // broken in old, works in new
+				Query: "select t1.i, t1.j, t2.i, t2.j from t1 natural right join t2;",
+				Expected: []sql.Row{
+					{nil, nil, 1, 30},
+					{2, 20, 2, 20},
+					{nil, nil, 5, 50},
+				},
+			},
+
+			// Nested Join
+			{
+				Query: "select t1.i, t1.j, t2.i, t2.j, t3.i, t3.j from t1 join t2 using (i) join t3 on t1.i = t3.i;",
+				Expected: []sql.Row{
+					{1, 10, 1, 30, 1, 200},
+					{2, 20, 2, 20, 2, 20},
+				},
+			},
+			{
+				Skip:        true, // broken in old, works in new
+				Query:       "select t1.i, t1.j, t2.i, t2.j, t3.i, t3.j from t1 join t2 on t1.i = t2.i join t3 using (i);",
+				ExpectedErr: sql.ErrAmbiguousColumnName,
+			},
+			{
+				Query: "select t1.i, t1.j, t2.i, t2.j, t3.i, t3.j from t1 join t2 using (i) join t3 using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1, 30, 1, 200},
+					{2, 20, 2, 20, 2, 20},
+				},
+			},
+			{
+				Query: "select * from t1 join t2 using (i) join t3 using (i);",
+				Expected: []sql.Row{
+					{1, 10, 30, 200},
+					{2, 20, 20, 20},
+				},
+			},
+
+			// Subquery Tests
+			{
+				Query: "select t1.i, t1.j, tt.i from t1 join (select 1 as i) tt using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1},
+				},
+			},
+			{
+				Query: "select t1.i, t1.j, tt.i, tt.j from t1 join (select * from t2) tt using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1, 30},
+					{2, 20, 2, 20},
+				},
+			},
+			{
+				Query: "select tt1.i, tt1.j, tt2.i, tt2.j from (select * from t1) tt1 join (select * from t2) tt2 using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1, 30},
+					{2, 20, 2, 20},
+				},
+			},
+
+			// CTE Tests
+			{
+				Query: "with cte as (select * from t1) select cte.i, cte.j, t2.i, t2.j from cte join t2 using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1, 30},
+					{2, 20, 2, 20},
+				},
+			},
+			{
+				Query: "with cte1 as (select * from t1), cte2 as (select * from t2) select cte1.i, cte1.j, cte2.i, cte2.j from cte1 join cte2 using (i);",
+				Expected: []sql.Row{
+					{1, 10, 1, 30},
+					{2, 20, 2, 20},
+				},
+			},
+			{
+				Query: "WITH cte(i, j) AS (SELECT 1, 1 UNION ALL SELECT i, j from t1) SELECT cte.i, cte.j, t2.i, t2.j from cte join t2 using (i);",
+				Expected: []sql.Row{
+					{1, 1, 1, 30},
+					{1, 10, 1, 30},
+					{2, 20, 2, 20},
+				},
+			},
+			{
+				Query: "with recursive cte(i, j) AS (select 1, 1 union all select i + 1, j * 10 from cte where i < 3) select cte.i, cte.j, t2.i, t2.j from cte join t2 using (i);",
+				Expected: []sql.Row{
+					{1, 1, 1, 30},
+					{2, 10, 2, 20},
+				},
+			},
+
+			// Broken CTE tests
```

Comment:
i'm seeing the same repro but i can't see why mysql would error on this/why it would be useful to error on this. will see if psql does the same on monday

Reply:
I'm not sure why MySQL would error on these either, must be some weird artifact of their analyzer the detects duplicate columns. We could just be more permissive than them here?